### ISSUE TITLE
CLI support?

### ISSUE THREAD

**Issue by halaxa** [halaxa/json-machine#73]

**Comment by fwolfsjaeger** [comment]
On Linux this can already be done using jq: https://stedolan.github.io/jq/

**Comment by halaxa** [comment]
Good point, I didn't know, that jq supports stream parsing. The speed will be incomparable, that's clear, but the jq's usage with stream parsing seems somehow unintuitive. While looking at the usage of jq another option came to my mind for `jm`:
```sh
$ wget <big list of users to stdout> | jm --pointer=/results
{"key": 0, "value": {"name": "Frank Sinatra", ...}}
{"key": 1, "value": {"name": "Ray Charles", ...}}
...
```
It is extensible for other fields in the future such as `position`, `matchedPointer` and so on...

**Comment by fwolfsjaeger** [comment]
You're right, using `jq` for stream parsing is not intuitive at all. A cli-wrapper for JSON Machine can be easily added:
```php
#!/usr/bin/env php
<?php

use JsonMachine\Items;

if ( ! is_file(dirname(__DIR__).'/vendor/autoload.php')) {
    throw new LogicException('Composer autoloader missing. Try running "composer install".');
}

require_once dirname(__DIR__).'/vendor/autoload.php';

function usage()
{
    echo sprintf('usage: %s --pointer=""', __FILE__)."\n";
    exit(1);
}

$options = getopt(null, ['pointer:']);

if (!isset($options['pointer'])) {
    usage();
}

$iterator = Items::fromFile('php://stdin', $options);

foreach ($iterator as $row) {
    echo json_encode($row)."\n";
}
```

**Comment by halaxa** [comment]
Yes, something along those lines. Using `PassthruDecoder` would eliminate the overhead of encoding and decoding back. Also, a simple templating system (Mustache for example) could be used to allow the user to format decoded data if that's what they wish. Like:
```sh
$ wget <big list of users to stdout> | jm --item-template="{{name}};{{address}}"
```
Combined with json pointer it could be quite versatile.

**Comment by pkoppstein** [comment]
For the uninitiated, jq's streaming parser is usually quite difficult to use, but worse, for the following two essential tasks (described here using standard jq syntax), it is typically very slow (many hours or days) for very big files:

1. `.[]` - that is, "explode" an array into a stream of its top-level items; 

2. `keys_unsorted[] as $k | {($k): .[$k]}` -- that is, "exploding" an object into a stream of corresponding singleton (i.e. single-key) objects.

To my knowledge, there is currently no CLI-tool for running these two jq queries conveniently, speedily, and losslessly against very large JSON arrays or objects, respectively.  (By "lossless" I mean avoiding the loss of precision in handling JSON numbers.)

Being able to use JSON Pointer to fine-tune the point of the "explosion" would be fantastic!

Thank you!

**Comment by pkoppstein** [comment]
@ fwolfsjaeger - Unfortunately your script does not preserve the JSON structure of the items at the specified point(s).

Or at least, I tried it with 'pointer' => '/-' and as input: `[1,2,[10,20],30]` but the array is in effect completely flattened.
--

Incidentally, after running `composer` successfully, I tried running your script (in the same directory), but the result is
a fatal error, with the message:

    Composer autoloader missing.

(In fact, both the files

./vendor/autoload.php
./vendor/halaxa/json-machine/src/autoloader.php

are present.)

**Comment by halaxa** [comment]
> Or at least, I tried it with 'pointer' => '/-' and with input: `[1,2,[10,20],30]` but the array is in effect completely flattened.

That's correct behavior. If you want to iterate top level, use empty string JSON pointer (default). Read more about it in REAMDE to see how exactly a hyphen in JSON pointer works. By using `/-` you tell it you want to iterate over each iterable on the top level array. But most of the values are scalars, so they are not iterated, just passed along as they are. But when it hits the only array you have there, it is itself iterated. So the result is flat.

**Comment by pkoppstein** [comment]
@halaxa - Thank you for your explanation.  Please understand that the difficulty I had was precisely because I read the README quite closely, the point being that in JSON, numbers are scalars, not iterables.  That is, I would have expected that an attempt to iterate over a number would either result in an error, or nothing at all.  (In jq, gojq, and jaq, it results in an error e.g. `$JQ -n '12 | .[]'`.)

Part of my confusion arose from statements such as the following in an "Overview of JSON Pointer": (*)

> Note that the first character of this String [the JSON Pointer] is a ‘/' – this is a syntactic requirement.

When I tried using "/" as the JSON Pointer, I just got an error, so I "/-" seemed like the next best bet.

The fact that #fwolfsjaeger's script requires a pointer didn't help my understanding. 

Now that I understand how to iterate over an array, I would like to know how to avoid loss of numeric precision, e.g.

400000000000000000000000000000000000000000000000000000000123 => 4.0e+59

Thank you again.

(*) https://www.baeldung.com/json-pointer

**Comment by halaxa** [comment]
No problem :)

This sencence
> Note that the first character of this String is a ‘/' – this is a syntactic requirement.

from here https://www.baeldung.com/json-pointer is **incorrect**. See https://www.rfc-editor.org/rfc/rfc6901#section-5. The official RFC is also linked from the JSON Machine README https://github.com/halaxa/json-machine#what-is-json-pointer-anyway.

`"/"` actually means *Iterate over empty string key in root dictionary*.

**Comment by halaxa** [comment]
I'll elaborate on the two other points of yours later. Hopefully tomorrow.

**Comment by pkoppstein** [comment]
@halaxa @fwolfsjaeger - My PHP was never good to begin with and is by now very rusty, but the following script has already proven useful to me and might provide a basis for further improvements.  Suggestions would of course be welcome.
```
#!/usr/bin/env php
<?php

# Author: peak@princeton.edu
# Author of JSON Machine: https://github.com/halaxa

$jmVersion="0.0.1 of 2022.10.10";

$bn=basename(__FILE__);

$dirs = array( __DIR__.'/vendor/autoload.php',
        dirname(__DIR__).'/vendor/autoload.php' );

$autoload = '';	
foreach ($dirs as &$value) {
  if (is_file($value)) { $autoload = $value; break; }
}   

if ($autoload == '') {
  $home = exec( 'echo ~' ) ;
  $autoload = $home . '/github/json-machine/src/autoloader.php';
  if (is_file($home)) {
     spl_autoload_register(require $autoload);
  } else {
     throw new LogicException('Composer autoloader missing. Try running "composer require halaxa/json-machine".');
  }
} else {
  require_once $autoload;
}  

use JsonMachine\Items;

function usage() {
  global $bn;
echo <<<END
Usage: {$bn} [ --pointer=JSONPOINTER ] [ FILEPATH ... ]
or:    {$bn} [-h | --help]
where JSONPOINTER defaults to '' and FILEPATH defaults to stdin.
\n
END;
}

function help() {
  global $bn, $jmVersion;  
echo <<<END
The -h option gives a usage summary, and --help gives more detailed help.

One of the main uses of {$bn} is to stream a JSON array or JSON object
that occurs at the top-level or within a very large JSON structure.

In this document, streaming a JSON array is to be understood as
producing a stream of the top-level items in the array (one line per
item), and similarly, streaming a JSON object means producing a stream
of the top-level values.  Streaming other JSON values simply means
printing them.

If no JSON Pointer is specified, or if it is specified as the empty
string (""), then the top-level entity is streamed.

If a JSON Pointer is specified, then all the corresponding entities
are streamed successively, as illustrated in Example (4) below.

Examples:
(1) jm <<< '[1,"2", {"a": 4}, [4]]'
yields:

1
"2"
{"a":4}
[4]

(2) jm <<< '{"a": 1, "b": [2,3]}'
yields

1
[2,3]

(3) jm --pointer "/results" <<< '{"results": {"a": 1, "b": [2,3]}}'
yields the same stream as (2) above.

(4) jm --pointer "/-" <<< '[1,[2,3]]'
yields

1
2
3

Note that in the last example, the JSON Pointer "/-" points in turn to 1 and [2,3],
and that streaming 1 produces 1, and streaming [2,3] produces 2 and then 3.

{$bn} is based on JSON Machine. For further details, see documentation on:

* "JSON Machine" e.g. https://github.com/halaxa/json-machine
* "JSON Pointer" e.g. https://www.rfc-editor.org/rfc/rfc6901#section-5

This is version {$jmVersion} of {$bn}.
\n
END;  
  exit(0);
}

$options = getopt('h', ['pointer:', 'help'], $rest_index);

if (isset($options['help']))  { usage(); help(); exit; }
if (isset($options['h']))     { usage(); exit; }

# if (isset($options['pointer'])) {
#   print "Note: JSONPOINTER is: " . $options['pointer'] . "\n";
# }

function process($filepath, $options) {
  $iterator = Items::fromFile( $filepath, $options);
  foreach ($iterator as $row) {
     echo json_encode($row)."\n";
  }
}

$pos_args = array_slice($argv, $rest_index);

$argx = count($pos_args);

if ($argx == 0) { process('php://stdin', $options); exit; }

# Interpret the remaining arguments as filepaths if possible:

foreach( $pos_args as &$filepath ) {
  if (! is_file($filepath)) {
    die(__FILE__ . ': specified file not found: ' . $filepath . "\n");
    exit(1);
  }
}

foreach( $pos_args as &$filepath ) {
  process( $filepath, $options );
}
```

**Comment by halaxa** [comment]
Can you move this last post to a new discussion, please?

**Comment by pkoppstein** [comment]
@halaxa - The last post is essentially a CLI script, so I thought this would be the best thread?

By the way, many of my colleagues who might benefit from a script such as jm would probably be discouraged by the installation hurdles that currently exist, so I was wondering whether you could envision at some point making JSON Machine available via `homebrew`?  Or are there other alternatives you can suggest?

**Comment by halaxa** [comment]
It sure is a cli script. But I understood you wanted some suggestions. Discussions would be better place for this. If you want to actually participate with some code to this repository, please use a pull request. This thread is mainly for ideas and suggestions about how should CLI interface work. As for other installation channels,  I'll let someone else to do it for now. It needs its own maintenance time which I don't have. It's OSS, anyone can generate any package from any revision. But thank you for your suggestion. Please keep them coming ;)

**Comment by pkoppstein** [comment]
@halaxa - I've created Issue#88 in accordance with your request.  Should I now delete the script from the message in this thread?

As mentioned in Issue#87, I'm not sure how JSON_BIGINT_AS_STRING helps, as it converts all "bigints" to strings, and doesn't even attempt to handle big or small decimals.

As for `SCALAR_ONLY`/`VECTOR_ONLY`, I think that the current behavior is fine, and maybe even the "correct" one, depending on how it's documented.  In the "help" provided by my version of the jm script, I've attempted to clarify the intent by emphasizing the idea of "streaming" rather than "iteration".

**Comment by halaxa** [comment]
> I've created Issue#88 in accordance with your request. Should I now delete the script from the message in this thread?

I agree with deleting it.

> As mentioned in Issue#87, I'm not sure how JSON_BIGINT_AS_STRING helps, as it converts all "bigints" to strings, and doesn't even attempt to handle big or small decimals.

Your example has an integer, so that's why I suggested it. What exact decimal problem are you facing?

**Comment by pkoppstein** [comment]
It's not clear to me how a CLI script can be implemented to handle a JSON files that contains more than one top-level JSON entity.  There are tools for converting such files to JSONLines format (one JSON entity per line), but it's inconvenient to have to place each of these in a separate file for the sake of JSON Machines.  In addition, some of these tools lose numerical precision. 

Any suggestions would be appreciated. Thanks.
